### DISCUSSION TITLE
CubatureGrid design and validation

### DISCUSSION THREAD

**Discussion by rafmudaf** [NatLabRockies/floris#883] [v4 Design Discussion]
# Grid to support cubature interpolation for rotor average velocity

FLORIS currently fully supports two types of grid: [TurbineGrid](https://github.com/NREL/floris/blob/8bbe62d820a2e98589cbe6fe688fbc877383e6b9/floris/core/grid.py#L108) and [FlowFieldPlanarGrid](https://github.com/NREL/floris/blob/8bbe62d820a2e98589cbe6fe688fbc877383e6b9/floris/core/grid.py#L496). Last year, a new type of grid was added in #649, [TurbineCubatureGrid](https://github.com/NREL/floris/blob/8bbe62d820a2e98589cbe6fe688fbc877383e6b9/floris/core/grid.py#L260), but the implementation and verification were incomplete.

## Background

The `TurbineGrid` arranges points in a rectangular layout on the rotor swept area for fast turbine power calculations, as shown below.
<img src="https://github.com/NREL/floris/assets/13797903/cd5bdf53-fc3a-4fb2-afb4-d5397dab8085" width="200">
The `FlowFieldPlanarGrid` places points in a regular (structured) grid pattern throughout the flow field to compute the wake in the space between turbines, as shown below. Note that using this type of grid requires first getting the velocities on a `TurbineGrid`.
<img src="https://github.com/NREL/floris/assets/13797903/4f2ab20d-487c-4af5-8e27-57c0497095ee" width="300">

Since the mathematical wake models in FLORIS depend on the thrust coefficient, an incoming wind speed must be assigned to each turbine. FLORIS does this by calculating the velocities at the `TurbineGrid` points and averaging them with a [cubic mean](https://github.com/NREL/floris/blob/8bbe62d820a2e98589cbe6fe688fbc877383e6b9/floris/core/rotor_velocity.py#L60). Given the coarse resolution and incomplete coverage of the rotor swept area of the `TurbineGrid` points, the shear profile and incoming wake profile are under sampled, and this leads to notable grid dependence (see #499).

To resolve this issue, the `TurbineCubatureGrid` along with the cubature integration method was added as an alternative to the cubic mean on a square grid for integrating values on a circle, as shown below.
<img src="https://github.com/NREL/floris/assets/13797903/d44127b9-21d0-4a69-bf7a-05a1cf8242cd" width="300">

## Implementation in FLORIS

The `TurbineCubatureGrid` implementation in FLORIS is incomplete and not validated, and the requirements to complete it are outlined here. 

During the transition to v4, this grid-type was partially updated, and it crashes in v4.0. This bug is fixed in #881. Another issue is reported in #709 where this grid-type crashes when used with the cumulative-curl model. This grid-type runs without error with all other models.

Remaining work:
- For odd values for order of integration (`N`), duplicate points are placed at the rotor center. Since the points each have the same weighting and locations, they are redundant and can be removed (with weighting adjusted accordingly) to improve efficiency
- Change the "grid_resolution" setting to "order" or something more meaningful to this grid-type
- Add tests for the grid class itself
- Add regression tests

Validation:
- Show that this method correctly integrates an analytical function over a circle
- Characterize the grid convergence (results as a function of `N`)
- Characterize the computational cost a function of `N` and relative to the `TurbineGrid`

Missing capabilities:
- There is no mechanism to specify that this grid-type should be used in the visualization functions (`calculate_*_plane`) since the `TurbineGrid` is hard-coded there.

**Comment by Darry6682** [comment]
![image](https://github.com/NREL/floris/assets/71394037/4b5c04ba-b628-42e0-b0bd-4be42fb3f3f6)
Hi @rafmudaf  Thanks for your great efforts making the FLORIS so powerful! I noticed that the TurbineGrid can only get the velocity at the rotor plane within the swept area, However, is there any method to obtain the nine-point profile for each WT as shown in this figure? This is important for my following research. Appreciate it if you could help!

**Reply by rafmudaf** [reply]
Yes, but you have to kind of work around some things. I'm happy to help you with it. Can you create a new discussion for that and tag me?
